osc/obs_api/xmlmodel/fields.py

```diff
@@ -0,0 +1,341 @@
+        model._root.tag = value
+
+    delete = None
+
+
+class AttributeField(Field):
+    def get(self, model):
+        result = model._root.attrib.get(self.name, None)
+        result = self._get_typed_value(result)
+        return result
+
+    def set(self, model, value):
+        for validator in self.validators:
+            validator(value)
+
+        if self.optional and value is None:
+            self.delete(model, validate=False)
+        else:
+            model._root.attrib[self.name] = self._set_typed_value(value)
+
+    def delete(self, model, validate=True):
+        if validate:
+            for validator in self.validators:
+                validator(None)
+
+        model._root.attrib.pop(self.name)
+
+
+class TextNodeField(Field):
+    def get(self, model):
+        node = model._root.find(self.name)
+
+        # return None if the node doesn't exist
+        if node is None:
+            return None
+
+        # return an empty string if the node exists and is empty
+        if node.text is None:
+            return ""
+
+        return node.text
+
+    def set(self, model, value):
+        for validator in self.validators:
+            validator(value)
+
+        self.delete(model, validate=False)
+
+        if value is None:
+            # just delete the element
+            return
+
+        # create new node based on ``value``
+        node = ET.SubElement(model._root, self.name)
+        node.text = value
+
+    def delete(self, model, validate=True):
+        if validate:
+            for validator in self.validators:
+                validator(None)
+
+        nodes = model._root.findall(self.name)
+        for node in nodes:
+            model._root.remove(node)
+
+
+class TextNodeListField(Field):
+    def get(self, model):
+        nodes = model._root.findall(self.name)
+        if nodes is None:
+            return ()
+        return tuple([i.text for i in nodes])
+
+    def set(self, model, values):
+        for value in values:
+            for validator in self.validators:
+                validator(value)
+
+        self.delete(model, validate=False)
+
+        # create new nodes based on ``values``
+        for value in values:
+            node = ET.SubElement(model._root, self.name)
+            node.text = value
+
+    def delete(self, model, validate=True):
+        if validate:
+            for validator in self.validators:
+                validator(None)
+
+        nodes = model._root.findall(self.name)
+        for node in nodes:
+            model._root.remove(node)
+
+    def validate(self, model, values, what=None):
+        if what:
+            what += f".{self.get_property_name(model)}"
+        else:
+            what = f"{self.get_property_name(model)}"
+
+        for validator in self.validators:
+            for value in values:
+                validator(value, what=what)
+
+
+class ModelField(Field):
+    def __init__(self, name, model_class, optional=False, help_text=None):
```

Comment:
Whether a field is optional should be obtainable from the type signature of the field. What I mean is that atm there's a bit of information duplication going on:
```Python
class Foo:
    trigger: Optional[str] = xmlmodel.AttributeField(
        "trigger",
        choices=choices.RELEASE_TRIGGERS,
        optional=True,
    )
```
We know that trigger is `optional=True` due to the type. Maybe it would be possible to hijack pydantic's validator syntax for this:
```Python
class Foo:
    @xmlmodel.AttributeField(choices.RELEASE_TRIGGERS)
    trigger: Optional[str]
```